### DISCUSSION TITLE
[2023-03-27] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6079] [Weekly Demo Day]
I was working on 2 main things this week:

1. [Execution Context Design](https://github.com/enso-org/design/pull/55). It required a lot of meetings and thinking in order to make it right. It bases on the work of @jdunkerley and it would not get done without tremendous help of @jdunkerley and @NedHarding, thank you so much ❤️.
2. I started working on the vector-editor widget. I found one issue which was preventing the graphical part from happening, namely, how we handle mouse events in the application. I created a [PR changing it to a way more flexible way](https://github.com/enso-org/enso/pull/6078).

Next week I will be mainly working on the visual part of the vector editor, hopefully delivering the first working version. This was planned for this week, but point 1 took me a lot of time that was not originally planned for my work this week.

**Comment by MichaelMauderer** [comment]
I was working on the UI part of the Execution Context Selection, as well as some preliminary controller integration. This will be review ready today, after finishing some integration and design refinement. 

https://user-images.githubusercontent.com/1428930/227940412-5e938fad-f684-48a6-9224-aeea657b9cd2.mp4

**Comment by Procrat** [comment]
Last week, I created some tooltips for the icon buttons next to nodes (and the profiling icon for consistency). If you notice that the text placement seems a bit off, there's work ongoing to fix that across all tooltips.

![](https://user-images.githubusercontent.com/607786/226972920-81033b37-001f-49eb-9fc6-453120f01760.gif)

**Comment by Frizi** [comment]
1. Last week I was working on integrating widgets into node view. So far I have managed to build a tree of individual text label widgets and display the code for each one of them. The text labels are laid out using automatic layouts, with which I had multiple issues. No interaction other than whole-node edit is handled yet.
2. Because the layout system wasn't working correctly, I was working on fixing its bugs together with @farmaazon, and we created a [PR with multiple bugfixes](https://github.com/enso-org/enso/pull/6066/). Also added support for alignment in manual layouts, but due to ongoing github outage I cannot create a PR for it yet.

Edit: I was able to push the branch and open a [PR for layout and sprite alignment](https://github.com/enso-org/enso/pull/6087).

![layout-anim-and-alignment](https://user-images.githubusercontent.com/919491/227951742-7a7fd48a-7d07-4e19-b824-8c136e3fb381.gif)

**Comment by farmaazon** [comment]
1. The raw vector editor is implemented and successfully integrated with the widget system. It reacts to code changes, but does not allow any editing: this will be continuously added in next tasks.

3. I used layout in the vector editor stub and spotted many issues in layout system. With @Frizi, we created a PR with fixes (as he mentioned above).

Besides layout problems, I have a problem with reacting to size changes, what makes the "jumping effect".

https://user-images.githubusercontent.com/3919101/227955735-f96fc23d-7e87-4042-8586-c1154523e871.mp4

**Comment by kazcw** [comment]
Last week I've been working on virtual sublayers (partitioning a symbol into different draw-orders, that can be drawn in one draw call). Implementation mainly involves adding support for moving symbol instances within a buffer (so I'm also implementing buffer compaction). I plan to finish tomorrow.

Next I will debug a performance problem on Intel Xe-LP. When that's fixed I'll work on reducing draw calls, using the superbox and virtual sublayers.

**Comment by mwu-tow** [comment]
Over the last week I was working on enabling easy project sharing and opening.

The result is https://github.com/enso-org/enso/pull/6077, currently fully functional and awaiting review.

Enso will now associate with two file extensions:
* `.enso` — Enso source file.
  * If the source file belongs to a project under the Project Manager-managed directory, it will be opened.
  * If the source file belongs to a project located elsewhere, it will be imported into the PM-managed directory and opened;
  * otherwise, opening the `.enseo` file will fail. (e.g, loose source file without any project)
* `.enso-project` — Enso project bundle, i.e. `tar.gz` archive containing a compressed Enso project directory.
  * it will be imported under PM-managed directory, if needed a unique directory name shall be generated.
